src/Shared/OpportunisticIntern.cs

```diff
@@ -21,116 +21,138 @@ namespace Microsoft.Build
     ///
     ///     string interned = OpportunisticIntern.Intern(String.Join(",",someStrings));
     ///
-    /// This class uses heuristics to decide whether it will be efficient to intern a string or not. There is no
+    /// There are currently two underlying implementations. The new default one in WeakStringCacheInterner is based on weak GC handles.
+    /// The legacy one in BucketedPrioritizedStringList is available only as an escape hatch by setting an environment variable.
+    ///
+    /// The legacy implementation uses heuristics to decide whether it will be efficient to intern a string or not. There is no
     /// guarantee that a string will intern.
     ///
     /// The thresholds and sizes were determined by experimentation to give the best number of bytes saved
     /// at reasonable elapsed time cost.
+    ///
+    /// The new implementation interns all strings but maintains only weak references so it doesn't keep the strings alive.
     /// </summary>
-    internal static class OpportunisticIntern
+    internal sealed class OpportunisticIntern
     {
-        private static readonly bool s_useSimpleConcurrency = Traits.Instance.UseSimpleInternConcurrency;
+        /// <summary>
+        /// Defines the interner interface as we currently implement more than one.
+        /// </summary>
+        private interface IInternerImplementation
+        {
+            /// <summary>
+            /// Converts the given internable candidate to its string representation. Efficient implementions have side-effects
+            /// of caching the results to end up with as few duplicates on the managed heap as practical.
+            /// </summary>
+            string InterningToString<T>(T candidate) where T : IInternable;
+
+            /// <summary>
+            /// Prints implementation specific interning statistics to the console.
+            /// </summary>
+            /// <param name="heading">A string identifying the interner in the output.</param>
+            void ReportStatistics(string heading);
+        }
+
+        /// <summary>
+        /// The singleton instance of OpportunisticIntern.
+        /// </summary>
+        private static OpportunisticIntern _instance = new OpportunisticIntern();
+        internal static OpportunisticIntern Instance => _instance;
+
+        private readonly bool _useLegacyInterner = Traits.Instance.UseLegacyStringInterner;
```

Comment:
Do you have to store this here rather than just using Traits.Instance.UseLegacyStringInterner a few times?

Reply:
If `Traits` explicitly guaranteed that the traits never change during execution it would make sense. I think I'd rather leave it like this (same as the pre-existing `_useSimpleConcurrency`) for consistency and a little bit of extra defensive safety.